<a href="https://colab.research.google.com/github/SteelworkerDB/Sales_Predictions/blob/main/Prediccion_Ventas.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **PROYECTO PREDICCIÓN DE VENTAS**

---




# CARGA DE DATOS Y VISUALIZACIÓN INICIAL

In [ ]:
import pandas as pd
filename="/content/sales_predictions_2023.csv"
vp_df=pd.read_csv(filename)
vp_df.head()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
0           FDA15         9.30          Low Fat             0.02   
1           DRC01         5.92          Regular             0.02   
2           FDN15        17.50          Low Fat             0.02   
3           FDX07        19.20          Regular             0.00   
4           NCD19         8.93          Low Fat             0.00   

               Item_Type  Item_MRP Outlet_Identifier  \
0                  Dairy    249.81            OUT049   
1            Soft Drinks     48.27            OUT018   
2                   Meat    141.62            OUT049   
3  Fruits and Vegetables    182.09            OUT010   
4              Household     53.86            OUT013   

   Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
0                       1999      Medium               Tier 1   
1                       2009      Medium               Tier 3   
2                       1999      Medium               Tier 1   
3                       1998         NaN               Tier 3   
4                       1987        High               Tier 3   

         Outlet_Type  Item_Outlet_Sales  
0  Supermarket Type1            3735.14  
1  Supermarket Type2             443.42  
2  Supermarket Type1            2097.27  
3      Grocery Store             732.38  
4  Supermarket Type1             994.71

In [ ]:
vp_df.tail()

Item_Identifier  Item_Weight Item_Fat_Content  Item_Visibility  \
8518           FDF22         6.87          Low Fat             0.06   
8519           FDS36         8.38          Regular             0.05   
8520           NCJ29        10.60          Low Fat             0.04   
8521           FDN46         7.21          Regular             0.15   
8522           DRG01        14.80          Low Fat             0.04   

               Item_Type  Item_MRP Outlet_Identifier  \
8518         Snack Foods    214.52            OUT013   
8519        Baking Goods    108.16            OUT045   
8520  Health and Hygiene     85.12            OUT035   
8521         Snack Foods    103.13            OUT018   
8522         Soft Drinks     75.47            OUT046   

      Outlet_Establishment_Year Outlet_Size Outlet_Location_Type  \
8518                       1987        High               Tier 3   
8519                       2002         NaN               Tier 2   
8520                       2004       Small               Tier 2   
8521                       2009      Medium               Tier 3   
8522                       1997       Small               Tier 1   

            Outlet_Type  Item_Outlet_Sales  
8518  Supermarket Type1            2778.38  
8519  Supermarket Type1             549.28  
8520  Supermarket Type1            1193.11  
8521  Supermarket Type2            1845.60  
8522  Supermarket Type1             765.67

# LIMPIEZA

---



In [ ]:
vp_df.isna().sum() #VERIFICAR CUANTOS DATOS ESTÁN EN BLANCO
#HAY 1463 DATOS EN NAN EN Item_Weight y 2410 en Outles_Size

Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

In [ ]:
vp_df.info() #PAR VERIFICAR EL TIPO DE DATOS Y EL NÚMERO DE COLUMNAS Y FILAS

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 8523 entries, 0 to 8522
Data columns (total 12 columns):
 #   Column                     Non-Null Count  Dtype  
---  ------                     --------------  -----  
 0   Item_Identifier            8523 non-null   object 
 1   Item_Weight                7060 non-null   float64
 2   Item_Fat_Content           8523 non-null   object 
 3   Item_Visibility            8523 non-null   float64
 4   Item_Type                  8523 non-null   object 
 5   Item_MRP                   8523 non-null   float64
 6   Outlet_Identifier          8523 non-null   object 
 7   Outlet_Establishment_Year  8523 non-null   int64  
 8   Outlet_Size                6113 non-null   object 
 9   Outlet_Location_Type       8523 non-null   object 
 10  Outlet_Type                8523 non-null   object 
 11  Item_Outlet_Sales          8523 non-null   float64
dtypes: float64(4), int64(1), object(7)
memory usage: 799.2+ KB


# 1. ¿CUÁNTAS FILAS Y COLUMNAS HAY?

---



COLUMNAS: 12
FILAS: 8523

# 2. ¿TIPOS DE DATOS DE CADA VARIABLE?

---



In [ ]:
vp_df.dtypes

Item_Identifier               object
Item_Weight                  float64
Item_Fat_Content              object
Item_Visibility              float64
Item_Type                     object
Item_MRP                     float64
Outlet_Identifier             object
Outlet_Establishment_Year      int64
Outlet_Size                   object
Outlet_Location_Type          object
Outlet_Type                   object
Item_Outlet_Sales            float64
dtype: object

# 3. ¿DUPLICADOS?

---



In [ ]:
vp_df.duplicated().sum() #VERIFICANDO DUPLICADOS

0

# 4. ¿VALORES FALTANTES?

---



In [ ]:
vp_df.isna().sum() #VERIFICAR CUANTOS DATOS ESTÁN EN BLANCO
#HAY 1463 DATOS EN NAN EN Item_Weight y 2410 en Outles_Size


Item_Identifier                 0
Item_Weight                  1463
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

# 5.  TRATAMIENTO DE VALORES FALTANTES
# 6. VERIFICACIÓN DE QUE NO HAYA VALORES FALTANTES DESPUÉS DE ABORDADOS

---




In [ ]:
#ANÁLISIS DE DATOS DE COLUMNA Item_Weight
#PARA DECIDIR TRATAMIENTO DE FALTANTES:


# AGRUPO SEGÚN COLUMNA Item_type PARA VERIFICAR SI EXISTEN DIFERENCIAS
# EN LAS DISTRIBUCIONES.
# ADICIONAL OBTENGO EL PROMEDIO TOTAL DE Item_Weight
prom_peso_tipo=vp_df.groupby('Item_Type')['Item_Weight'].mean().reset_index()
std_peso_tipo=vp_df.groupby('Item_Type')['Item_Weight'].std().reset_index()
prom_peso_total=vp_df['Item_Weight'].mean()
print('PROMEDIO DE PESOS POR TIPO\n', prom_peso_tipo)
print()
print('DEVIACIÓN ESTÁNDAR DE PESOS POR TIPO\n',std_peso_tipo)
print()
print('PROMEDIO TOTAL DE Item_Weight\n',prom_peso_total)

PROMEDIO DE PESOS POR TIPO
                 Item_Type  Item_Weight
0            Baking Goods        12.28
1                  Breads        11.35
2               Breakfast        12.77
3                  Canned        12.31
4                   Dairy        13.43
5            Frozen Foods        12.87
6   Fruits and Vegetables        13.22
7             Hard Drinks        11.40
8      Health and Hygiene        13.14
9               Household        13.38
10                   Meat        12.82
11                 Others        13.85
12                Seafood        12.55
13            Snack Foods        12.99
14            Soft Drinks        11.85
15          Starchy Foods        13.69

DEVIACIÓN ESTÁNDAR DE PESOS POR TIPO
                 Item_Type  Item_Weight
0            Baking Goods         4.77
1                  Breads         4.44
2               Breakfast         5.04
3                  Canned         4.59
4                   Dairy         4.69
5            Frozen Foods         4.

 **OBSERVACIONES: Las distribuciones parecen ser parecidas.  El promedio total de pesos no se desvía mucho
 de los promedios de peso por tipo de ítem.**

In [ ]:
#PARA VISUALIZAR SI EXISTE UNA RELACIÓN ENTRE LOS VALORES FALTANTES
#DE PESO Y LOS VALORES DE OTRA COLUMNA, COMO LA DE Item_Type
#SUMO LA CANTIDAD DE FALTANTES

faltantes_tipo=vp_df.groupby('Item_Type',group_keys=True)['Item_Weight'].apply(lambda x:x.isna().sum())
faltantes_tipo

#ES DIFICIL ENCONTRAR ALGUNA RELACIÓN DE LA INFORMACIÓN OBSERVADA CON LOS
#DATOS FALTANTES,.  LOS FALTANTES PARECEN SER MCAR

Item_Type
Baking Goods             112
Breads                    47
Breakfast                 21
Canned                   110
Dairy                    116
Frozen Foods             138
Fruits and Vegetables    213
Hard Drinks               31
Health and Hygiene        90
Household                151
Meat                      88
Others                    32
Seafood                   13
Snack Foods              212
Soft Drinks               71
Starchy Foods             18
Name: Item_Weight, dtype: int64

**OBSERVACIONES: No parece haber relación entre el tipo de ítem y los valores faltantes de peso.**

**LOS VALORES FALTANTES DE Item_weight PARECEN SER FALTANTES COMPLETAMENTE ALEATORIOS POR LO MENOS AL ANALIZAR LOS VALORES AGRUPADOS DE Item_Type e Item_Weight.**  

**ADEMÁS, EL PROMEDIO DE LOS PESOS PUEDE IMPUTARSE DENTRO DE LOS VALORES FALTANTES, POR LO MENOS SIN SESGO SIGNIFICATIVO HACIA EL TIPO DE ÍTEM.**


In [ ]:
#IMPUTACIÓN: Reemplazo de valores faltantes de Item_Weight con el promedio
vp_df_imp=vp_df
vp_df['Item_Weight']=vp_df['Item_Weight'].fillna(prom_peso_total)
vp_df_imp.isna().sum() #VERIFICAR CUANTOS DATOS ESTÁN EN BLANCO


Item_Identifier                 0
Item_Weight                     0
Item_Fat_Content                0
Item_Visibility                 0
Item_Type                       0
Item_MRP                        0
Outlet_Identifier               0
Outlet_Establishment_Year       0
Outlet_Size                  2410
Outlet_Location_Type            0
Outlet_Type                     0
Item_Outlet_Sales               0
dtype: int64

**SE VERIFICÓ QUE NO HAY FALTANTES EN Item_Weight**

In [ ]:
#PARA EL CASO DE LA COLUMNA OUTLET SIZE, EXISTEN TRES POSIBLES VALORES.
#VERIFICO CUÁNTAS VECES SE REPITEN LOS DATOS EN LA COLUMNA PARA
#PLANIFICAR UNA ESTRATEGIA DE IMPUTACIÓN.

vp_df_imp['Outlet_Size'].value_counts()

Medium    2793
Small     2388
High       932
Name: Outlet_Size, dtype: int64

In [ ]:
#DEBIDO A QUE EL NUMERO DE FALTANTES ES COMPARABLE A LA CANTIDAD
#QUE EXISTE DE CADA VALOR: MEDIUM, SMALL, HIGH,
#SERÍA CONVENIENTE UNA ESTRATEGIA QUE DISTRIBUYA LOS FALTANTES ENTRE CADA UNO
#DE ESTOS VALORES.  NOTA: NO ES BUENA IDEA ELIMINAR LAS FILAS PORQUE ES
#UNA CANTIDAD MUY ALTA DE VALORES FALTANTES.

#UNA ESTRATEGIA ADECUADA SERÍA EL FORWARD FILL, LEER LA OBSERVACIÓN
#INMEDIATAMENTE ANTERIOR Y REEMPLAZAR A LA SIGUIENTE.
#SE USA EL MÉTODO FFILL DE FILLNA

vp_df_imp['Outlet_Size'].fillna(method='ffill', inplace=True)

#VERIFICANDO QUE NO HAYA FALTANTES:
vp_df_imp.isna().sum()

Item_Identifier              0
Item_Weight                  0
Item_Fat_Content             0
Item_Visibility              0
Item_Type                    0
Item_MRP                     0
Outlet_Identifier            0
Outlet_Establishment_Year    0
Outlet_Size                  0
Outlet_Location_Type         0
Outlet_Type                  0
Item_Outlet_Sales            0
dtype: int64

# 7.  ARREGLANDO CATEGORÍA INCONSISTENTE DE DATOS


---



In [ ]:
vp_df_imp.groupby('Item_Fat_Content')['Item_Identifier'].count().reset_index()


Item_Fat_Content  Item_Identifier
0               LF              316
1          Low Fat             5089
2          Regular             2889
3          low fat              112
4              reg              117

OBSERVANDO LOS DATOS LA COLUMNA Item_Fat_Content CONTIENE 3 FORMAS DE INDICAR LOW FAT (Low Fat, LF, low fat) Y 2 FORMAS DE INDICAR REGULAR (reg y Regular)

In [ ]:
#PARA RESOLVER LA INCONSISTENCIA, REEMPLAZO LOS VALORES DE reg CON Regular y
#LOS VALORES DE LF con Low Fat.

vp_df_imp['Item_Fat_Content'].replace(['LF','low fat','reg'],
                                        ['Low Fat', 'Low Fat', 'Regular'],
                                      inplace=True)
#VERIFICANDO QUE SE RESOLVIÓ LA INCONSISTENCIA CORRECTAMENTE
vp_df_imp.groupby('Item_Fat_Content')['Item_Identifier'].count().reset_index()


Item_Fat_Content  Item_Identifier
0          Low Fat             5517
1          Regular             3006

# 8. PARA CADA COLUMNA NUMÉRICA, OBTENER LAS ESTADÍSTICAS RESUMIDAS

---



In [ ]:
#SIMPLEMENTE UTILIZANDO LA FUNCIÓN DESCRIBE

vp_df_imp.describe().round(2)



Item_Weight  Item_Visibility  Item_MRP  Outlet_Establishment_Year  \
count      8523.00          8523.00   8523.00                    8523.00   
mean         12.86             0.07    140.99                    1997.83   
std           4.23             0.05     62.28                       8.37   
min           4.56             0.00     31.29                    1985.00   
25%           9.31             0.03     93.83                    1987.00   
50%          12.86             0.05    143.01                    1999.00   
75%          16.00             0.09    185.64                    2004.00   
max          21.35             0.33    266.89                    2009.00   

       Item_Outlet_Sales  
count            8523.00  
mean             2181.29  
std              1706.50  
min                33.29  
25%               834.25  
50%              1794.33  
75%              3101.30  
max             13086.96